### ISSUE TITLE
Oddity in match default case

### ISSUE THREAD

**Issue by StevenTschantz** [coq/coq#16138]
I've discovered what to me is another oddity in match.  Apparently it doesn't know what to do (or at least what I mean to do) when there is only a default case to match against.  I can't imagine what alternative meaning there would be as a deliberate design decision.  This is an edge case to be sure.  From deep (for me) in the weeds of an overly complex development, I offer the following simplified example.  The last definition fails unexpectedly.  I'm using all variations of explicit match terms because I haven't seemed to figure out how to get standard tactics that work on the dependently typed structures I'm working with.  I'm still on 8.13.2 on Windows 10.

```
Require Import Numbers.BinNums.

Inductive testI : positive -> Type :=
| testI_xI : forall (p1 : positive), testI (xI p1)
| testI_xO : forall (p1 : positive), testI (xO p1)
(* | testI_xH : *never* *)
.

Definition testI_not_xH_good (x : testI xH) : False :=
match x as x0 in testI p0 return
  match p0
  with
  | xH => False
  | _ => True
  end
with
| testI_xI p1 => I
| testI_xO p1 => I
end.

Definition testI_not_xH_also_good (x : testI xH) : False :=
match x as x0 in testI p0 return
  match p0
  with
  | xH => False
  | _ => True
  end
with
| testI_xI p1 => I
| _ => I
end.

Definition testI_not_xH_bad (x : testI xH) : False :=
match x as x0 in testI p0 return
  match p0
  with
  | xH => False
  | _ => True
  end
with
| _ => I
end.
```

- Steven Tschantz

**Comment by mrhaandi** [comment]
Maybe this is a feature (not a bug). We have
```coq
(* this is ok *)
Definition is_eq (x : unit) : 
  match x with
  | _ => True
  end = True := eq_refl.

(* this fails *)
Fail Definition is_not_eq (x : unit) : 
  match x with
  | tt => True
  end = True := eq_refl.
```
So a match with a single underscore pattern is in fact not a match at all.
As soon as you have at least one non-underscore pattern, it is in fact a match term.

In your example 
```coq
with
| testI_xI p1 => I
| _ => I
end.
```
is the correct term for the match, whereas 
```coq
with
| _ => I
end
```
is a different (not definitionally equal) term which has no match.


As a result (also a minimal example for your problem):
```coq
(* this is ok *)
Definition test_ok (x : unit) :
  match x with
  | tt => True
  end :=
  match x with
  | tt => I
  end.

(* this is ok *)
Definition test_ok2 (x : unit) :
  match x with
  | _ => True
  end :=
  match x with
  | _ => I
  end.

(* this fails, also tested with Coq 8.16+rc1 *)
Fail Definition test_fails (x : unit) :
  match x with
  | tt => True
  end :=
  match x with
  | _ => I
  end.
```

**Comment by mrhaandi** [comment]
By the way, in your special case the following works
```coq
Definition testI_not_xH (x : testI xH) : False.
Proof.
  inversion x.
Defined. (* or Qed. in case of Lemma *)
```
However, inversion is unreliable for dependent terms. There is the theory of `small inversions` under active development for that.

**Comment by StevenTschantz** [comment]
I agree that a match with a single underscore pattern turns out to be interpreted as a no match at all term.  The is_eq example shows the difference in a way that might be useful, but I would argue that if you go to the trouble of writing a match at all, you must have wanted a match to be carried out, as in my example, to some purpose.  This is of course not an instance of some sort of match eta rule, only of whether the initial input is interpreted as a match term at all.  Now perhaps I can see that a general pattern matching capability with arguments of constructors also being constructors to some depth implying nested matches might be seen as having a limiting case where no constructors at all are mentioned as being a depth zero match (maybe making a recursive approach to arbitrary depth easier).  But I'm not sure this is the design choice I would have made.  It's merely a minor annoyance to get around this, now that I know what went wrong (in my understanding).

Thanks for the comments about inversion, confirming at least that I'm not overlooking something easy.  I'm working on an excessively complex (for me) dependent types development while struggling to figure out the basics of a) what works conveniently; b) what can be made to work; and most importantly, c) what can't possibly work because I have a fundamental misunderstanding.  I offer this by way of explanation as to why it is that I'm writing so many of my own match terms that I'm stumbling over such oddities.

- Steven Tschantz

**Comment by Janno** [comment]
I agree that the interpretation of `match .. with _ end` is very surprising. It should not be the case that spelling out the possible constructors with `_` for all arguments is any different from using just `_` (or a variable) for the pattern. In particular, this code here
```coq
Definition testI_not_xH_bad (x : testI xH) : False :=
match x as x0 in testI p0 return
  match p0 with
  | xH => False
  | _ => True
  end
with
| (testI_xI _ | testI_xO _) => I
end.
```
should be exactly the same as 
```coq
Definition testI_not_xH_bad (x : testI xH) : False :=
match x as x0 in testI p0 return
  match p0 with
  | xH => False
  | _ => True
  end
with
| _ => I
end.
```

Now, I know that there is a pattern that people use where they employ such a "non-match" to create an evar in a particular scope in a way that requires *no* reduction. This is best described by the difference between `bla1` and `bla2` below.
```coq
Notation "'bla1' T1 ,  T2" := (fun y : T1 => (_ : T2)) (at level 0).
Check bla1 nat , bool.
(*
fun y : nat => ?b : bool
     : nat -> bool
where
?b : [y : nat |- bool]
*)
Notation "'bla2' T1 ,  T2" := (match _ with x => fun y : T1 => (x : T2) end) (at level 0).
Check bla2 nat , bool.
(*
fun _ : nat => ?b : bool
     : nat -> bool
where
?b : [ |- bool]
*)
```
Note that `x`'s scope in `bla2` does not include `y` *and*, crucially, that the term is identical to `fun y => _` (modulo the difference in scopes) after interpretation, i.e.  no reduction is necessary to get rid of the match term.

I personally work on developments where this pattern is used and I am mentioning it because fixing the counter-intuitive behavior described in this issue will possible require a new way to control an evar's scope.
